<a href="https://colab.research.google.com/github/Sar-mon/Ragastone12/blob/main/04_%2C_DataFrame_ipynb_Sara_Moncada.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **0. Libreries**

In [49]:
import pandas as pd    ## Manage the data base
import numpy as np     ## Algebraic operations
import plotly.express as px    ## Graphics

# **1. Import the dataset**

In [56]:
# Save the file path
p = "/content/sample_data/Data_Sara_DV.xlsx"

# Read and save the dataset
d = pd.read_excel(p)

# **2. Basic Dataframe exploration**

In [57]:
# Revizar la información general del dataframe

d.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 231 entries, 0 to 230
Data columns (total 8 columns):
 #   Column                                                                                Non-Null Count  Dtype 
---  ------                                                                                --------------  ----- 
 0   Country Name                                                                          231 non-null    object
 1   Country Code                                                                          231 non-null    object
 2   Time                                                                                  231 non-null    int64 
 3   Time Code                                                                             231 non-null    object
 4   GDP growth (annual %) [NY.GDP.MKTP.KD.ZG]                                             231 non-null    object
 5   Trade (% of GDP) [NE.TRD.GNFS.ZS]                                                     231 non

In [58]:
# View the first 5 rows (Muesta las primeras 5 filas del dataset, dentro del parentesis se puede poner el numero de filas para visualizar)

d.head(5)

Country Name Country Code  Time Time Code  \
0   Costa Rica          CRI  2013    YR2013   
1   Costa Rica          CRI  2014    YR2014   
2   Costa Rica          CRI  2015    YR2015   
3   Costa Rica          CRI  2016    YR2016   
4   Costa Rica          CRI  2017    YR2017   

  GDP growth (annual %) [NY.GDP.MKTP.KD.ZG] Trade (% of GDP) [NE.TRD.GNFS.ZS]  \
0                                  2.494766                         64.066411   
1                                   3.54211                         65.191242   
2                                  3.652081                         60.753562   
3                                  4.204323                         62.178478   
4                                  4.157699                         65.071954   

   Net migration [SM.POP.NETM]  \
0                         3950   
1                         3996   
2                         4065   
3                         4150   
4                         4196   

  Unemployment, total (% of total labor force) (national estimate) [SL.UEM.TOTL.NE.ZS]  
0                                              8.218                                    
1                                              8.499                                    
2                                              8.415                                    
3                                              8.085                                    
4                                              7.729

In [59]:
# View the last 5 rows (Muestra las ultimas 5 filas del dataframe)

d.tail(15)

Country Name Country Code  Time Time Code  \
216  St. Vincent and the Grenadines          VCT  2020    YR2020   
217  St. Vincent and the Grenadines          VCT  2021    YR2021   
218  St. Vincent and the Grenadines          VCT  2022    YR2022   
219  St. Vincent and the Grenadines          VCT  2023    YR2023   
220             Trinidad and Tobago          TTO  2013    YR2013   
221             Trinidad and Tobago          TTO  2014    YR2014   
222             Trinidad and Tobago          TTO  2015    YR2015   
223             Trinidad and Tobago          TTO  2016    YR2016   
224             Trinidad and Tobago          TTO  2017    YR2017   
225             Trinidad and Tobago          TTO  2018    YR2018   
226             Trinidad and Tobago          TTO  2019    YR2019   
227             Trinidad and Tobago          TTO  2020    YR2020   
228             Trinidad and Tobago          TTO  2021    YR2021   
229             Trinidad and Tobago          TTO  2022    YR2022   
230             Trinidad and Tobago          TTO  2023    YR2023   

    GDP growth (annual %) [NY.GDP.MKTP.KD.ZG]  \
216                                 -3.739146   
217                                  0.753797   
218                                  7.159833   
219                                  6.016109   
220                                  3.531787   
221                                  3.853635   
222                                  -0.77746   
223                                 -7.527137   
224                                 -4.800828   
225                                 -0.602465   
226                                  0.361272   
227                                 -9.077877   
228                                 -1.037426   
229                                  1.482235   
230                                  2.103245   

    Trade (% of GDP) [NE.TRD.GNFS.ZS]  Net migration [SM.POP.NETM]  \
216                                ..                         -379   
217                                ..                         -379   
218                                ..                         -200   
219                                ..                         -200   
220                                ..                         -796   
221                                ..                         -799   
222                                ..                         -583   
223                                ..                         -103   
224                                ..                         1446   
225                                ..                        35502   
226                                ..                       -19248   
227                                ..                         2370   
228                                ..                         2632   
229                                ..                         -800   
230                                ..                         -800   

    Unemployment, total (% of total labor force) (national estimate) [SL.UEM.TOTL.NE.ZS]  
216                                                 ..                                    
217                                                 ..                                    
218                                                 ..                                    
219                                                 ..                                    
220                                              2.683                                    
221                                              2.477                                    
222                                              2.428                                    
223                                              3.207                                    
224                                                 ..                                    
225                                                 ..                                    
226                        

# **3. Cleaning**

## **3.1. Column names**

In [60]:
# Change GDP growth (annual %) [NY.GDP.MKTP.KD.ZG] to GDP Growth (anual %)
d.rename(columns = {'GDP growth (annual %) [NY.GDP.MKTP.KD.ZG]':'GDP Growth (anual %)'}, inplace = True)

# Change Trade (% of GDP) [NE.TRD.GNFS.ZS] to Trade (% of GDP)
d.rename(columns = {'Trade (% of GDP) [NE.TRD.GNFS.ZS]':'Trade (% of GDP)'}, inplace = True)

# Change Net migration [SM.POP.NETM] to Net Migration
d.rename(columns = {'Net migration [SM.POP.NETM]':'Net Migration'}, inplace = True)

# Change Unemployment, total (% of total labor force) (national estimate)
# [SL.UEM.TOTL.NE.ZS] to Total Unemployment
d.rename(columns = {'Unemployment, total (% of total labor force) (national estimate) [SL.UEM.TOTL.NE.ZS]':'Total Unemployment'}, inplace = True)



In [62]:
d.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 231 entries, 0 to 230
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Country Name          231 non-null    object
 1   Country Code          231 non-null    object
 2   Time                  231 non-null    int64 
 3   Time Code             231 non-null    object
 4   GDP Growth (anual %)  231 non-null    object
 5   Trade (% of GDP)      231 non-null    object
 6   Net Migration         231 non-null    int64 
 7   Total Unemployment    231 non-null    object
dtypes: int64(2), object(6)
memory usage: 14.6+ KB


In [63]:
# Duplicate the dataset (Se duplica en caso de algun error tener la base original)
d2 = d

In [64]:

# Create a new column with year as STR
# Check if 'year' column exists before accessing it
if 'year' in d2.columns:
    d2['year_str'] = d2['year'].astype(str)
else:
    print("Column 'year' not found in the DataFrame.")
    # If 'year' column doesn't exist, you need to investigate why and potentially add it
    # or modify your code accordingly

d2.info()

Column 'year' not found in the DataFrame.
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 231 entries, 0 to 230
Data columns (total 8 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Country Name          231 non-null    object
 1   Country Code          231 non-null    object
 2   Time                  231 non-null    int64 
 3   Time Code             231 non-null    object
 4   GDP Growth (anual %)  231 non-null    object
 5   Trade (% of GDP)      231 non-null    object
 6   Net Migration         231 non-null    int64 
 7   Total Unemployment    231 non-null    object
dtypes: int64(2), object(6)
memory usage: 14.6+ KB


## **3.2. Delete columns**

In [65]:
d2 = d2.drop(columns = ['Country Code'])

In [66]:
d2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 231 entries, 0 to 230
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Country Name          231 non-null    object
 1   Time                  231 non-null    int64 
 2   Time Code             231 non-null    object
 3   GDP Growth (anual %)  231 non-null    object
 4   Trade (% of GDP)      231 non-null    object
 5   Net Migration         231 non-null    int64 
 6   Total Unemployment    231 non-null    object
dtypes: int64(2), object(5)
memory usage: 12.8+ KB


## **3.3. Delete NaN**

In [67]:
# Replace by NaN
d2 = d2.replace('..', np.nan)
d2.tail(15)

<ipython-input-67-abac6c943964>:2: FutureWarning:

Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`



Country Name  Time Time Code  GDP Growth (anual %)  \
216  St. Vincent and the Grenadines  2020    YR2020             -3.739146   
217  St. Vincent and the Grenadines  2021    YR2021              0.753797   
218  St. Vincent and the Grenadines  2022    YR2022              7.159833   
219  St. Vincent and the Grenadines  2023    YR2023              6.016109   
220             Trinidad and Tobago  2013    YR2013              3.531787   
221             Trinidad and Tobago  2014    YR2014              3.853635   
222             Trinidad and Tobago  2015    YR2015             -0.777460   
223             Trinidad and Tobago  2016    YR2016             -7.527137   
224             Trinidad and Tobago  2017    YR2017             -4.800828   
225             Trinidad and Tobago  2018    YR2018             -0.602465   
226             Trinidad and Tobago  2019    YR2019              0.361272   
227             Trinidad and Tobago  2020    YR2020             -9.077877   
228             Trinidad and Tobago  2021    YR2021             -1.037426   
229             Trinidad and Tobago  2022    YR2022              1.482235   
230             Trinidad and Tobago  2023    YR2023              2.103245   

     Trade (% of GDP)  Net Migration  Total Unemployment  
216               NaN           -379                 NaN  
217               NaN           -379                 NaN  
218               NaN           -200                 NaN  
219               NaN           -200                 NaN  
220               NaN           -796               2.683  
221               NaN           -799               2.477  
222               NaN           -583               2.428  
223               NaN           -103               3.207  
224               NaN           1446                 NaN  
225               NaN          35502                 NaN  
226               NaN         -19248               3.523  
227               NaN           2370               4.206  
228               NaN           2632               4.446  
229               NaN           -800               4.383  
230               NaN           -800                 NaN

In [68]:
d2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 231 entries, 0 to 230
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Country Name          231 non-null    object 
 1   Time                  231 non-null    int64  
 2   Time Code             231 non-null    object 
 3   GDP Growth (anual %)  230 non-null    float64
 4   Trade (% of GDP)      141 non-null    float64
 5   Net Migration         231 non-null    int64  
 6   Total Unemployment    128 non-null    float64
dtypes: float64(3), int64(2), object(2)
memory usage: 12.8+ KB


In [69]:
# Drop NaN
d2 = d2.dropna()
d2.info()

<class 'pandas.core.frame.DataFrame'>
Index: 97 entries, 0 to 171
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Country Name          97 non-null     object 
 1   Time                  97 non-null     int64  
 2   Time Code             97 non-null     object 
 3   GDP Growth (anual %)  97 non-null     float64
 4   Trade (% of GDP)      97 non-null     float64
 5   Net Migration         97 non-null     int64  
 6   Total Unemployment    97 non-null     float64
dtypes: float64(3), int64(2), object(2)
memory usage: 6.1+ KB


In [70]:
d2.tail(15)

Country Name  Time Time Code  GDP Growth (anual %)  \
124  Dominican Republic  2016    YR2016              6.659200   
125  Dominican Republic  2017    YR2017              4.666705   
126  Dominican Republic  2018    YR2018              6.982528   
127  Dominican Republic  2019    YR2019              5.052168   
128  Dominican Republic  2020    YR2020             -6.720239   
129  Dominican Republic  2021    YR2021             12.271990   
130  Dominican Republic  2022    YR2022              4.858324   
131  Dominican Republic  2023    YR2023              2.360616   
165             Jamaica  2013    YR2013              0.517686   
166             Jamaica  2014    YR2014              0.689822   
167             Jamaica  2015    YR2015              0.921487   
168             Jamaica  2016    YR2016              1.375223   
169             Jamaica  2017    YR2017              0.997123   
170             Jamaica  2018    YR2018              1.889924   
171             Jamaica  2019    YR2019              0.892292   

     Trade (% of GDP)  Net Migration  Total Unemployment  
124         51.585971         -30181               7.284  
125         50.225085         -30055               5.832  
126         52.055966         -29944               5.861  
127         51.011145         -29830               6.356  
128         44.293657         -14966               6.126  
129         52.734867         -14966               7.702  
130         54.231526         -29294               5.504  
131         49.615332         -29099               5.555  
165         83.259572         -11329               9.698  
166         84.740999         -11329               9.046  
167         76.117522         -11329              13.511  
168         76.455202         -11329               8.634  
169         83.523173         -11329               7.399  
170         89.977906         -11329               5.470  
171         90.110608         -11329               4.987

# **4. Indexing**

Categories of the dataset (Categorizar las columnas principales)

In [71]:
# Indexing
d2 = d2.set_index(['Country Name', 'Time'])
d2.head(15)

Time Code  GDP Growth (anual %)  Trade (% of GDP)  \
Country Name Time                                                     
Costa Rica   2013    YR2013              2.494766         64.066411   
             2014    YR2014              3.542110         65.191242   
             2015    YR2015              3.652081         60.753562   
             2016    YR2016              4.204323         62.178478   
             2017    YR2017              4.157699         65.071954   
             2018    YR2018              2.615904         66.945342   
             2019    YR2019              2.417512         65.760530   
             2020    YR2020             -4.273354         60.139886   
             2021    YR2021              7.935762         70.781642   
             2022    YR2022              4.551492         79.659195   
             2023    YR2023              5.111922         70.286815   
El Salvador  2013    YR2013              2.182952         80.451194   
             2014    YR2014              1.729400         78.104426   
             2015    YR2015              2.400000         76.560271   
             2016    YR2016              2.539062         72.819093   

                   Net Migration  Total Unemployment  
Country Name Time                                     
Costa Rica   2013           3950               8.218  
             2014           3996               8.499  
             2015           4065               8.415  
             2016           4150               8.085  
             2017           4196               7.729  
             2018           4198               9.010  
             2019           4154              10.807  
             2020           2084              16.432  
             2021           2084              15.139  
             2022           3850              11.325  
             2023           3750               8.325  
El Salvador  2013         -49086               3.693  
             2014         -49130               4.154  
             2015         -51282               3.998  
             2016         -47499               4.418

In [72]:
d2.info()

<class 'pandas.core.frame.DataFrame'>
MultiIndex: 97 entries, ('Costa Rica', 2013) to ('Jamaica', 2019)
Data columns (total 5 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Time Code             97 non-null     object 
 1   GDP Growth (anual %)  97 non-null     float64
 2   Trade (% of GDP)      97 non-null     float64
 3   Net Migration         97 non-null     int64  
 4   Total Unemployment    97 non-null     float64
dtypes: float64(3), int64(1), object(1)
memory usage: 4.8+ KB


# **5. Descriptive statistics**

In [73]:
d2.describe()

GDP Growth (anual %)  Trade (% of GDP)  Net Migration  \
count             97.000000         97.000000      97.000000   
mean               3.085210         75.811743  -14087.391753   
std                4.233067         23.276083   18709.007418   
min              -17.668333         26.222849  -58095.000000   
25%                2.182952         55.496423  -27343.000000   
50%                3.484485         76.204442  -11329.000000   
75%                4.666705         93.984983    1283.000000   
max               15.835765        137.634074   11130.000000   

       Total Unemployment  
count           97.000000  
mean             6.604186  
std              3.488192  
min              1.700000  
25%              4.154000  
50%              5.861000  
75%              8.415000  
max             18.548000

In [74]:
# Formatear mejor los datos para el comando d2.describe() a dos decimales.
d2.describe().apply(lambda s: s.apply('{0:.2f}'.format))

GDP Growth (anual %) Trade (% of GDP) Net Migration Total Unemployment
count                97.00            97.00         97.00              97.00
mean                  3.09            75.81     -14087.39               6.60
std                   4.23            23.28      18709.01               3.49
min                 -17.67            26.22     -58095.00               1.70
25%                   2.18            55.50     -27343.00               4.15
50%                   3.48            76.20     -11329.00               5.86
75%                   4.67            93.98       1283.00               8.41
max                  15.84           137.63      11130.00              18.55

In [75]:
# Formatear los datos para el comando d2.describe() separar los miles con coma.
d2.describe().apply(lambda s: s.apply('{:,.2f}'.format))

GDP Growth (anual %) Trade (% of GDP) Net Migration Total Unemployment
count                97.00            97.00         97.00              97.00
mean                  3.09            75.81    -14,087.39               6.60
std                   4.23            23.28     18,709.01               3.49
min                 -17.67            26.22    -58,095.00               1.70
25%                   2.18            55.50    -27,343.00               4.15
50%                   3.48            76.20    -11,329.00               5.86
75%                   4.67            93.98      1,283.00               8.41
max                  15.84           137.63     11,130.00              18.55

*Analize* the data...


* The mean for GDP growth is 3.41% and the mean for trade (as a percentage of GDP) is 82.78%.
* The median for GDP growth is 3.53% and the median for trade (as a percentage of GDP) is 76.79%.

* The standard deviation of all numeric variables can be seen in the third row of the output of d2.describe(). For example, the standard deviation of GDP growth is 4.26% and the standard deviation of trade (as a percentage of GDP) is 50.93%.
* The minimum value of all numeric variables can be seen in the fourth row of the output of d2.describe(). For example, the minimum value of GDP growth is -11.65% and the minimum value of trade (as a percentage of GDP) is 16.21%.
* The maximum value of all numeric variables can be seen in the seventh row of the output of d2.describe(). For example, the maximum value of GDP growth is 14.77% and the maximum value of trade (as a percentage of GDP) is 285.46%.

# 7. Data Visualization

## 7.1 Line Graph

### i want a line grapah for gdp growth for Costa Rica, Where the x-axis is using plotly


In [92]:

import plotly.express as px


In [96]:
# Filter data for Costa Rica
costa_rica_df = d[d['Country Name'] == 'Costa Rica']

# Create the line graph
# Changed 'GDP Growth' to 'GDP Growth (anual %)' in the y argument
fig = px.line(costa_rica_df,
              x='Time',
              y='GDP Growth (anual %)',
              title='GDP Growth in Costa Rica')

fig.show()

In [79]:
costa_rica_data.head()

Time Code  GDP Growth (anual %)  Trade (% of GDP)  Net Migration  \
Time                                                                    
2013    YR2013              2.494766         64.066411           3950   
2014    YR2014              3.542110         65.191242           3996   
2015    YR2015              3.652081         60.753562           4065   
2016    YR2016              4.204323         62.178478           4150   
2017    YR2017              4.157699         65.071954           4196   

      Total Unemployment  
Time                      
2013               8.218  
2014               8.499  
2015               8.415  
2016               8.085  
2017               7.729

In [98]:
# Assuming your DataFrame is named 'd' and has columns 'Country Name', 'Time', and a value column (e.g., 'GDP Growth (anual %)')

# Filter data for the desired countries
countries = ['Costa Rica', 'El Salvador', 'Trinidad and Tobago']
filtered_df = d[d['Country Name'].isin(countries)]

# Create the line graph
# Replace 'GDP Growth (anual %)' with the actual column you want to plot on the y-axis
fig = px.line(filtered_df,
              x='Time',
              y='GDP Growth (anual %)',  # Changed 'Value' to 'GDP Growth (anual %)'
              color='Country Name',
              title='Line Graph for Costa Rica, El Salvador, and Trinidad and Tobago')

fig.show()